### PR TITLE
fix #3165 if round half up behavior is desired

### LINKED ISSUE #3165: jax.scipy.ndimage.map_coordinates round half is different than scipy
`jax.scipy.ndimage.map_coordinates` implements 'round half to even' when using nearest-neighbor interpolation and the coordinat values are exactly half (i.e. both 1.5 and 2.5 get rounded to 2). This is the default behavior for jax and ordinary numpy rounding (round, around, rint). However scipy.ndimage.map_coordinates appears to implement 'round half up'.

The difference between jax and scipy is illustrated with the following commands.
`>>> x = np.arange(10).astype(np.float32)`
`>>> coords = [[3.5, 4.5, 5.5, 6.5, 7.5]]`
`>>> jax.scipy.ndimage.map_coordinates(x, coords, order=0)`
`DeviceArray([4., 4., 6., 6., 8.], dtype=float32)`
`>>> scipy.ndimage.map_coordinates(x, coords, order=0)`
`array([4., 5., 6., 7., 8.], dtype=float32)`

Not sure which behavior is desired, but if the default numpy behavior is desired, the jax.scipy documentation should mention this. If the scipy behavior is desired, the rounding function at 
[https://github.com/google/jax/blob/7d157c71815e5386f2b449bac69ef9e3ab7e701f/jax/scipy/ndimage.py#L40](https://github.com/google/jax/blob/7d157c71815e5386f2b449bac69ef9e3ab7e701f/jax/scipy/ndimage.py#L40)
should be changed to something like:
`index = jnp.floor(coordinate + .5).astype(jnp.int32)`

### PR DESCRIPTION
**Pull request by jenkspt**

### CODE REVIEW

**Comment by jenkspt** [comment]
See [https://github.com/google/jax/issues/3165](https://github.com/google/jax/issues/3165)

**Comment by shoyer** [comment]
Good catch!

Would you mind adding a unit test as well to verify this change? See `tests/scipy_ndimage_test.py`

**Comment by jenkspt** [comment]
Also appears incorrect for integer `input`

`>>> x = np.arange(10)    # Integers`
`>>> jax.scipy.ndimage.map_coordinates(x, [[3.25, 3.5, 3.75, 4.25, 4.5]], order=0)`
`DeviceArray([3, 3, 3, 4, 4], dtype=int32)`
`>>> scipy.ndimage.map_coordinates(x, [[3.25, 3.5, 3.75, 4.25, 4.5]], order=0)`
`array([3, 4, 4, 4, 5])`

This is due to:
[https://github.com/google/jax/blob/f349979302548a9b1e93796875ce01b49f187eb8/jax/scipy/ndimage.py#L56](https://github.com/google/jax/blob/f349979302548a9b1e93796875ce01b49f187eb8/jax/scipy/ndimage.py#L56)

which should be:
`coordinates = [jnp.asarray(c) for c in coordinates]`

**Comment by jenkspt** [comment]
The addition of integer dtypes to the test case is what's causing the fails. I'll have to look into this some more.

File: jax/scipy/ndimage.py [hunk 1]
```diff
@@ -100,7 +100,9 @@ def _map_coordinates(input, coordinates, order, mode, cval):
       contribution = jnp.where(all_valid, input[indices], cval)
     outputs.append(_nonempty_prod(weights) * contribution)
   result = _nonempty_sum(outputs)
-  return result
+  if jnp.issubdtype(input.dtype, jnp.integer):
+    result = jnp.around(result)
+  return result.astype(input.dtype)
```
Comment (shoyer) [inline jax/scipy/ndimage.py:105]: This behavior for integer valued dtypes is not what I would have guessed for non-nearest neighbor interpolation! But I guess we don’t have any good reason to deviate from SciPy’s behavior here.

Comment (jenkspt) [reply jax/scipy/ndimage.py:105]: Ok -- you're right to be skeptical. Round half up matches scipy here -- need to flesh out the test again to catch this case as well.

File: jax/scipy/ndimage.py [hunk 2]
```diff
@@ -37,7 +37,7 @@
 
 
 def _nearest_indices_and_weights(coordinate):
-  index = jnp.around(coordinate).astype(jnp.int32)
+  index = jnp.floor(coordinate + .5).astype(jnp.int32)
```
Comment (shoyer) [inline jax/scipy/ndimage.py:40]: Could you define a helper function for this, with a name like `_round_half_up`? That would make the intent of this code more obvious (and maybe be useful below as well).

Comment (jenkspt) [reply jax/scipy/ndimage.py:40]: This is a rabbit hole. Scipy rounding is symmetric which means -1.5 and 1.5 get rounded to -2 and 2 respectively. but `np.floor(-1.5 + .5) == -1`

Comment (jenkspt) [reply jax/scipy/ndimage.py]: done

File: jax/scipy/ndimage.py (hunk 1 above)
Comment (jenkspt) [reply jax/scipy/ndimage.py:113]: fixed

**Comment by shoyer** [comment]
If round away from zero is what you want, just use lax.round which implement those semantics directly in XLA.

**Comment by shoyer** [comment]
It probably makes no difference in speed to use lax.round but it’s less code to maintain

**Comment by jenkspt** [comment]
lax.round doesn't take integer arguments. Would you prefer I keep the helper function like this?:
```python
def _round_half_away_from_zero(a):
  return a if jnp.issubdtype(a.dtype, jnp.integer) else lax.round(a)
```

**Comment by shoyer** [comment]
Sure, that would work!

**Comment by jenkspt** [comment]
Done! Hopefully ready to merge once the build passes.

**Comment by shoyer** [comment]
Thanks @jenkspt !

**Pull request merged**